### PR TITLE
Fixed detection of properties with underscores on nested forms

### LINKED ISSUE #3763: GetElementAccessPath in AdminHelper gets confused when using nested properties with underlines
| Question       | Answer
|----------------|-------------------------------------------------------------
| Bundle version | dev-master
| Symfony version| 2.8.5
| php version    | 5.5.34

This is my first issue on Sonata so if I am doing something wrong or you need more info I can provide it. I can also provide a PR that fixes this.

# Error message

Could not get element id from s5729b4b1cced5_main_category_current_tags Failing part: tags

# Stack trace

    [1] Exception: Could not get element id from s5729bd3e0bfe6_main_category_current_tags Failing part: tags
    at n/a in /vagrant/app/cache/dev/classes.php line 4358
    at Sonata\AdminBundle\Admin\AdminHelper->getElementAccessPath('s5729bd3e0bfe6_main_category_current_tags', object(Post)) in /vagrant/app/cache/dev/classes.php line 4264
    at Sonata\AdminBundle\Admin\AdminHelper->appendFormFieldElement(object(PostAdmin), object(Post), 's5729bd3e0bfe6_main_category_current_tags') in /vagrant/vendor/sonata-project/admin-bundle/Controller/HelperController.php line 105
    at Sonata\AdminBundle\Controller\HelperController->appendFormFieldElementAction(object(Request)) in  line 
    at call_user_func_array(array(object(HelperController), 'appendFormFieldElementAction'), array(object(Request))) in /vagrant/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php line 139
    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), '1') in /vagrant/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php line 62
    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), '1', true) in /vagrant/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/DependencyInjection/ContainerAwareHttpKernel.php line 69
    at Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel->handle(object(Request), '1', true) in /vagrant/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php line 185
    at Symfony\Component\HttpKernel\Kernel->handle(object(Request)) in /vagrant/web/app_dev.php line 30

# Steps to reproduce

Have 3 models: Post, Category, Tag. Post with a one-to-one with Category and Category with a one-to-many with Tag.

In the sonata configuration have 'sonata_type_model' on Category in PostAdmin and 'sonata_type_collection' on Tag in the CategoryAdmin.

Then try to add Tags from the PostAdmin

# Expected results

You should be able to create tags on the category, both from the CategoryAdmin and PostAdmin.

# Actual results

If you are using properties like: main_category and current_tags you can't.

### PR DESCRIPTION
**Pull request by jordisala1991**
Fixes #3763

Reopened with the correct base branch.

With this fix, getElementAccessPath now detects better when a property is accesible and shows better exception message when it's not. Added PHPDocs on the AdminHelperTest and fixed the compatibility with symfony 2.3.

I added more test to be sure it's working.

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Please fill your PR of missing parts indicated by the "Needs" labels.

Feel free to ask if you want some precision of what is needed.

**Comment by jordisala1991** [comment]
Can you explain a little bit?

Should I change CHANGELOG.md and add under the Unreleased / Added section an explanation of the PR?

**Comment by Soullivaneuh** [comment]
> Should I change CHANGELOG.md and add under the Unreleased / Added section an explanation of the PR?

Yes, but it should be on the `Fixed` section.

More information: http://keepachangelog.com/

**Comment by jordisala1991** [comment]
Added the changelog change

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 ### Added
+### Fixed
```
Comment (Soullivaneuh) [inline CHANGELOG.md]: Could you please add a blank line before `### Fixed`?

Comment (jordisala1991) [inline CHANGELOG.md]: Done

File: Admin/AdminHelper.php [hunk 2]
```diff
@@ -335,6 +334,8 @@ private function pathExists(PropertyAccessorInterface $propertyAccessor, $entity
             return true;
         } catch (NoSuchPropertyException $e) {
             return false;
+        } catch (UnexpectedTypeException $e) {
```
Comment (Soullivaneuh) [inline Admin/AdminHelper.php]: Why this?

Comment (jordisala1991) [inline Admin/AdminHelper.php]: When the adminHelper is trying to guess the correct path given a string it could end calling the method getValue of propertyAccessor passing as a parameter a thing different to an object or array. This is only for sf2.3.

It happened on my test and that's why it was failing on the compatibility test. I could instead change the test and remove that catch. But I thought it was more robust this way.

Comment (Soullivaneuh) [inline Admin/AdminHelper.php]: > This is only for sf2.3.

In this case, could you add a comment saying since which SF version we can remove this and why? Thanks.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Added
 
+### Fixed
+- Fixed detection of path when using nested properties in ``AdminHelper:getElementAccessPath`` method
```
Comment (Soullivaneuh) [inline CHANGELOG.md]: "Fix"

File: Admin/AdminHelper.php (hunk 2 above)
Comment (jordisala1991) [inline Admin/AdminHelper.php]: // Symfony <= 2.3 did not have isReadable method for PropertyAccessor
    if (method_exists($propertyAccessor, 'isReadable')) {
        return $propertyAccessor->isReadable($entity, $path);
    }

    try {
        $propertyAccessor->getValue($entity, $path);

        return true;
    } catch (NoSuchPropertyException $e) {
        return false;
    } catch (UnexpectedTypeException $e) {
        return false;
    }

It is already commented in the function. Should I add more information?

File: CHANGELOG.md (hunk 3 above)
Comment (jordisala1991) [inline CHANGELOG.md]: Done

File: Admin/AdminHelper.php [hunk 4]
```diff
@@ -284,13 +285,11 @@ public function getElementAccessPath($elementId, $entity)
...
                     $dot = '.';
                     $id = '';
-                } else {
-                    $dot = '';
                 }
             }
 
-            if ($dot !== '.') {
-                throw new \Exception(sprintf('Could not get element id from %s Failing part: %s', $elementId, $subValue));
+            if (!empty($id)) {
```
Comment (Soullivaneuh) [inline Admin/AdminHelper.php]: I'm sorry but I want to be sure. Why did you change this condition and the `else` below?

I'll try to reproduce your issue on 3.x and apply your issue next. :+1:

Comment (jordisala1991) [inline Admin/AdminHelper.php]: No problem, I will try to explain.

The else was just for throwing the exception and it's the part of the code that makes the issue appear.

And I changed the condition to be consistent with the exception. When the id is not empty is because it failed at finding the right path, and the part failing is exactly id, not subValue like before.

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Added
 
+### Fixed
+- Fix detection of path when using nested properties in ``AdminHelper:getElementAccessPath`` method
```
Comment (Soullivaneuh) [inline CHANGELOG.md]: I have a better understanding of your issue since: https://github.com/sonata-project/SonataAdminBundle/issues/3763#issuecomment-217808893

So this fix issue for underscore cased properties. Could you please add this precision?

**Comment by jordisala1991** [comment]
Do I need to provide more information about this issue @Soullivaneuh ?

**Comment by OskarStark** [comment]
@jordisala1991 can you please provide a good commit message and squash your commits?

A change of the PR title would be good too! Thank you!

**Comment by OskarStark** [comment]
@greg0ire can you please have a look at this PR? thank you!

**Comment by Soullivaneuh** [comment]
@OskarStark I still need to try to reproduce the issue. #3763

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Added
 
+### Fixed
+- Fix detection of path when using nested properties with underscores in ``AdminHelper:getElementAccessPath`` method
```
Comment (Soullivaneuh) [inline CHANGELOG.md]: Only one backtick please.

**Comment by greg0ire** [comment]
The README show how to squash your commits, please have a look at it.

File: Admin/AdminHelper.php (hunk 4 above)
Comment (greg0ire) [inline Admin/AdminHelper.php]: Why not throw the exception directly in the else part?

File: CHANGELOG.md (hunk 6 above)
Comment (OskarStark) [inline CHANGELOG.md]: good catch @Soullivaneuh ;-)

File: Tests/Admin/AdminHelperTest.php [hunk 7]
```diff
@@ -110,27 +98,38 @@ public function testAddNewInstanceInflector()
...
+
+        $path = $this->helper->getElementAccessPath('uniquePartOfId_path_to_object_0_another_more_things', $object);
+
+        $this->assertSame('path_to_object[0].another.more_things', $path);
+    }
 
+    /**
+     * @expectedException        Exception
+     * @expectdExceptionCode     0
+     * @expectedExceptionMessage Could not get element id from uniquePartOfId_path_to_object_0_more_calls Failing part: calls
```
Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: Please use the `setExpectedException` instead. It is better because you can't make mistake like line 121 (there is a typo)

File: Admin/AdminHelper.php (hunk 4 above)
Comment (jordisala1991) [inline Admin/AdminHelper.php]: Because if you throw the exception on the else part, you stop looking for attributes matching.

For example, if your attribute is called more_part. The foreach first look at more and it goes to the 'else' part. So if you throw an exception there it will stop looking at 'parts'. Without the else it will look for more_parts in the next iteration

**Comment by jordisala1991** [comment]
I will fix everything tonight, thanks for the input!

**Comment by OskarStark** [comment]
Thanks for you effort 👍

File: Admin/AdminHelper.php (hunk 4 above)
Comment (greg0ire) [inline Admin/AdminHelper.php]: So if I understand well, it only fails when and only when the last iteration fails, am I right ?

Comment (jordisala1991) [inline Admin/AdminHelper.php]: More or less...

The strings passed to this methods looks like: SomeId_attribute1_attribute2_attribute3part1_attribute3part2 ...etc

This issue is all about using the same separator '_' can be used both for the property separator and in a property name. This method can't guess automagically how the property is called so it will try to match all the string. Basically is trying to translate things like this: attribute1_attribute2_attribute3part1_attribute3part2 in attribute1.attribute2.attribute3part1_attribute3part2

So yes, the last iteration must go well, because if not the id will be 'not empty' and then will throw an exception.

Comment (greg0ire) [inline Admin/AdminHelper.php]: That's a great explanation, could you add it in a code comment, so that it is not lost? Or maybe make the code more explicit about that? It's hard to understand…

Comment (jordisala1991) [inline Admin/AdminHelper.php]: I prefer not to add comments to the code, but if necessary I will do it...

First I will try to make the code more self explanatory.
My first intention was not to change a lot of code in that function because otherwise it would be harder to review but it's true that the code is a bit hard to understand.

Comment (greg0ire) [inline Admin/AdminHelper.php]: > I prefer not to add comments to the code, but if necessary I will do it...

You're probably referring to the saying that goes something like this:

> If you need to comment your code, you failed to make it self-explanatory

In this instance, the code was failed to be made self-explanatory AND the author failed to admit his failure and make a comment.

So starting to make the code more self-explanatory is indeed the way to go, but in the end, if you feel a comment would not hurt, please don't refrain from adding one.

Comment (greg0ire) [inline Admin/AdminHelper.php]: A good starting point could be using a meaningful boolean : 
```php
if ($lastPathFindingSucceeded = $this->pathExists(
    $propertyAccessor, 
    $currentEntity, 
    $partReturnValue.$dot.$id
)) {
}
```

File: Admin/AdminHelper.php [hunk 8]
```diff
@@ -284,13 +285,11 @@ public function getElementAccessPath($elementId, $entity)
                     $partReturnValue .= $dot.$id;
                     $dot = '.';
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: Is there still a need for this to be variable ? Its value is now `''`, then `'.'` forever, maybe a `SEPARATOR` constant could be used, or simply `'.'` ?

File: Admin/AdminHelper.php [hunk 9]
```diff
@@ -284,13 +285,11 @@ public function getElementAccessPath($elementId, $entity)
                     $partReturnValue .= $dot.$id;
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: Instead of this concatenation, a `$detectedPathsParts` array could be used, and then imploded with `.`

**Comment by jordisala1991** [comment]
What do you think about the new implementation @greg0ire ?

File: Admin/AdminHelper.php [hunk 10]
```diff
@@ -254,53 +255,28 @@ public function getElementAccessPath($elementId, $entity)
...
-
-            if (isset($parts[$key + 1])) {
-                $currentEntity = $propertyAccessor->getValue($entity, $returnValue);
-            }
+        if (!empty($currentPath)) {
+            throw new \Exception(sprintf('Could not get element id from %s Failing part: %s', $elementId, $currentPath));
         }
 
-        return $returnValue;
+        return $totalPath;
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: `returnValue` as to be the worst variable name after `data`, thanks for changing it!

File: Admin/AdminHelper.php [hunk 11]
```diff
@@ -254,53 +255,28 @@ public function getElementAccessPath($elementId, $entity)
...
-                    $dot = '.';
-                    $id = '';
-                } else {
-                    $dot = '';
-                }
+            if ($this->pathExists($propertyAccessor, $entity, $totalPath.$separator.$currentPath)) {
+                $totalPath .= $separator.$currentPath;
+                $currentPath = '';
             }
+        }
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: Well this is way more simple like this! Thanks a lot!

**Comment by greg0ire** [comment]
Great work @jordisala1991 !

Please read and apply the CONTRIBUTING.md

- make one commit per thing you change
- don't make any commit to fix things in the same PR, squash them with the commits that introduced the problem

File: Admin/AdminHelper.php (hunk 11 above)
Comment (OskarStark) [inline Admin/AdminHelper.php]: good job 👍

File: Tests/Admin/AdminHelperTest.php [hunk 12]
```diff
@@ -18,13 +18,21 @@
 
 class AdminHelperTest extends \PHPUnit_Framework_TestCase
 {
-    public function testGetChildFormBuilder()
+    /**
+     * @var AdminHelper
+     */
+    protected $helper;
```
Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: should we do this change in a separate PR?

File: Tests/Admin/AdminHelperTest.php [hunk 13]
```diff
@@ -110,41 +98,40 @@ public function testAddNewInstanceInflector()
...
 
-        $this->assertSame('path_to_object[0].and.more', $path);
+        $this->helper->getElementAccessPath($path, $object);
     }
 
-    /**
-     * tests only so far that actual value/object is retrieved.
-     *
-     * @expectedException        Exception
-     * @expectedExceptionCode    0
```
Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: i would keep the `0` in the `setExpectedException` call

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: Why? It is the default value.

File: Admin/AdminHelper.php (hunk 2 above)
Comment (greg0ire) [inline Admin/AdminHelper.php]: To be removed when dropping sf 2.3, am I right ? Which is planned for version 4.0 of the bundle.

File: Tests/Admin/AdminHelperTest.php (hunk 13 above)
Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: ok

File: Tests/Admin/AdminHelperTest.php (hunk 12 above)
Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: At least in a separate commit.

File: Admin/AdminHelper.php (hunk 2 above)
Comment (jordisala1991) [inline Admin/AdminHelper.php]: Yes, that method should be removed once sf2.3 support is dropped.

**Comment by jordisala1991** [comment]
So I should squash all commits related to the same purpose... or something.

Like, in this case I should have 2 commits, one for the new implementation and test, and another for the refactor on the tests (creation of the setUp function and move the repeated code there). Is this correct?

**Comment by greg0ire** [comment]
> Is this correct?

Yes :) Good job!

File: Admin/AdminHelper.php (hunk 2 above)
Comment (OskarStark) [inline Admin/AdminHelper.php]: @jordisala1991 can you please create a PR after we merged this one, where you remove this stuff, and explain WHY we remove it and refer to this PR?

Thank you 🍺

**Comment by OskarStark** [comment]
Thank you for your effort on this PR :heavy_check_mark: 👍

**Comment by Soullivaneuh** [comment]
I would like to have issue **confirmed** before merge this. (https://github.com/sonata-project/SonataAdminBundle/issues/3763#issuecomment-217810605)

Did anyone of @sonata-project/contributors reproduced it?

**Comment by OskarStark** [comment]
No, but he reworked a method and added tests which are still green. IMO IF it fixes additional problems fine, otherwise the old stuff is working and the code is refactored to be more readable...

**Comment by Soullivaneuh** [comment]
@OskarStark If the issue cannot be reproduces, it open possibility to add useless code handling something we don't have to handle (if not related for example).

This is important to reproduce the bug before to be sure of what we do. Would be better to follow that next time.

I will try to add some labels to make it clearer.

Please keep also in mind that code coverage of this project is not good at all and green tests does not mean everything is ok.

**Comment by OskarStark** [comment]
> I will try to add some labels to make it clearer.

Thanks, will check it twice the next time 👍 

> Please keep also in mind that code coverage of this project is not good at all and green tests does not mean everything is ok.

Indeed, thats true!

File: Admin/AdminHelper.php (hunk 2 above)
Comment (jordisala1991) [inline Admin/AdminHelper.php]: Should I do it on the master branch? This can only be removed if sf2.3 support is dropped. @OskarStark

**Comment by jordisala1991** [comment]
I will provide a full functional example with a github repository of the issue with the sonata admin without this PR. Give me some time to prepare things. @Soullivaneuh
